### PR TITLE
Detectcone cleanup.patch

### PR DESCRIPTION
**Pull request by bjornborg**
This temporarily locks tiny-dnn version to latest known to be working one. There is also an issue that xtensor implementation in tiny-dnn doesnt comply with gcc8. I had to lock docker base image from alpine:edge to alpine:3.8.

Also added minor patches to source code to avoid compilation errors, as well did some restructuring and cleanup in the repo.

I did not test the functionality of the module.

What do you think about the changes?

### CODE REVIEW

**Comment by arbulgazar** [comment]
I think it's fine. I will work later or downloading `cluon` properly if we decide to do it that way. Should I merge it or you want someone else to review as well?

**Comment by bjornborg** [comment]
I think it would be healthy if one more could take a look. (Free exp)

File: CMakeLists.txt [hunk 1]
```diff
@@ -122,8 +123,8 @@ include_directories(SYSTEM ${LIBRT_INCLUDE_DIR})
 
 include_directories(SYSTEM /usr/include)
 include_directories(SYSTEM /usr/local/include)
-link_directories(SYSTEM /usr/lib) #Access g2o
-link_directories(SYSTEM /usr/local/lib) #Access g2o
+# link_directories(SYSTEM /usr/lib) #Access g2o
```
Comment (arbulgazar) [inline CMakeLists.txt]: What is this g2o?

Comment (bjornborg) [reply CMakeLists.txt]: I believe it's junk leftover from copying the dockerfile from other repo. If I remember correctly, g2o was a third party software used in slamORB filtering.

Comment (bjornborg) [reply CMakeLists.txt]: They can be safely removed if you want.

Comment (arbulgazar) [reply CMakeLists.txt]: Would prefer to remove them then. I can do that after the merge though. There is still much to be done in cleanup effort anyway.

Comment (bjornborg) [reply CMakeLists.txt]: Yeah, you can do it after the merge.

**Review by ruguangyou — COMMENTED** [review]
I think it's fine as well, compilation is okay. would start to think about testing, and set some test plans tomorrow

File: Dockerfile.amd64 [hunk 2]
```diff
@@ -30,30 +30,40 @@ RUN apk update && \
...
+        libtbb@testing \
+        libtbb-dev@testing \
+        openblas \
+        openblas-dev
 
 WORKDIR /tmp
-ADD ./config.h /opt/sources/config.h
+# ADD ./include/config.h /opt/sources/config.h
 
+# Why do we have this if we have a similar cmake.Modules folder? Should we have a similar approach?
```
Comment (chrberger) [inline Dockerfile.amd64]: This git pull to copy only one single is somewhat over-kill.

File: Dockerfile.amd64 [hunk 3]
```diff
@@ -30,30 +30,40 @@ RUN apk update && \
...
    mkdir -p eigen-git-mirror/build && \
    cd eigen-git-mirror/build && \
    cmake .. && \
    make install
-RUN echo "@testing http://nl.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories
-RUN apk update && \
-    apk --no-cache add \
-opencv@testing
+
+RUN git clone https://github.com/tiny-dnn/tiny-dnn.git && \
```
Comment (chrberger) [inline Dockerfile.amd64]: Here, you should only clone the latest master instead of the entire project history (--depth 1)

File: Dockerfile.amd64 [hunk 4]
```diff
@@ -70,6 +80,8 @@ MAINTAINER Christian Berger "christian.berger@gu.se"
 
 #Start microservice
 WORKDIR /usr/bin
+# This will make the image heavy... Do we need all of them?
+# This could be slimmed down
 COPY --from=builder /usr/local/include/ /usr/include/
```
Comment (chrberger) [inline Dockerfile.amd64]: Why do we need the include headers for the runtime binary?

File: Dockerfile.amd64 [hunk 5]
```diff
@@ -70,6 +80,8 @@ MAINTAINER Christian Berger "christian.berger@gu.se"
...
 COPY --from=builder /usr/local/lib/ /usr/lib/
```
Comment (chrberger) [inline Dockerfile.amd64]: For the libs: you can run `ldd` to see what the dependent shared libraries are to carry over.

**Review by chrberger — CHANGES_REQUESTED** [review]

**Comment by namv1994** [comment]
The compilation is (temporarily) fixed for now. Currently I'm fiixing some errors in forward detection, will take a look at Christian's reviews in the weekend.

File: Dockerfile.amd64 (hunk 2 above)
Comment (bjornborg) [reply Dockerfile.amd64]: Turns out that the file wasn't even needed.

File: Dockerfile.amd64 (hunk 5 above)
Comment (bjornborg) [reply Dockerfile.amd64]: Agreed, I'll leave this as homework for the cfsd team.

**Comment by ruguangyou** [comment]
After successfully building the docker image, I ran the command "docker run -ti detectcone:test /bin/sh", (the image is named detectcone:test in my case), and got inside the container. Then I got following results:
image
The program showed segmentation fault somehow.

**Comment by bjornborg** [comment]
I suspected that upx binary compression might cause this crash and successfully identified it to be the culprit. I've had same problem before where the upx compression causes the segmentation fault. Do you know more about this @chrberger? I've disabled the upx for now.

**Comment by chrberger** [comment]
@bjornborg upx can be tricky when we are using the wrong version/OS combination. I have also sometimes experienced non-working combinations. As we are not depending on it, disabling it is a good workaround for now.

File: Dockerfile.amd64 [hunk 6]
```diff
@@ -76,7 +78,81 @@ RUN apk update && \
 WORKDIR /usr/bin
 # This will make the image heavy... Do we need all of them?
 # This could be slimmed down
-COPY --from=builder /usr/local/lib/ /usr/lib/
-COPY --from=builder /usr/lib/ /usr/lib/
+# Nothing in the builder's /usr/local/lib
```
Comment (chrberger) [inline Dockerfile.amd64:81]: I like this change!

Comment (bjornborg) [reply Dockerfile.amd64:81]: :+1:

**Pull request merged**